Implemented NCCL Distributed Backend for PyTorch with new dist APIs

teng-li opened the pull request:
This PR added a new PyTorch distributed backend called NcclDataChannel that uses NVIDIA NCCL2.0+ for collective operations across different GPUs on different nodes.  The Nccl backend uses TCP socket as the rendevous for the initial broadcast of the unique NCCL ID.

The following new PyTorch APIs are added to perform collective operations on multiple GPU tensors, each of which resides on a different GPU.

```
    torch.distributed.all_reduce_multigpu(tensor_list, op=reduce_op.SUM, group=group.WORLD)
    torch.distributed.reduce_multigpu(tensor_list, dst, op=reduce_op.SUM, group=group.WORLD)
    torch.distributed.all_gather_multigpu(output_tensor_lists, input_tensor_list, group=group.WORLD)
    torch.distributed.broadcast_multigpu(tensor_list, src, group=group.WORLD)
    torch.distributed.destroy_process_group()
```

**How to use?**

```
torch.distributed.init_process_group("nccl")
# Optional
grp = torch.distributed.new_group()
torch.distributed.all_reduce_multigpu(tensor_list, group=grp)
# Clean shutdown to release all GPU resources
torch.distributed.destroy_process_group()
```

The exiting pytorch API such as torch.distributed.all_reduce works with "NCCL" backend as well.
Added some tiny new code paths for distributed data parallel model to use NCCL backend.

**Testing:**

All new functions are tested using my own test scripts. 

On DGX1, 8 nodes with 4 Infinibands. I am able to run ResNet50 (that uses the modified distributed parallel model) for 140+ epochs without any issues.

**TODO** will be writing unit-test for the entire backend.  Current marked as experimental only

teng-li committed:
Implemented NCCL Distributed Backend for PyTorch with new dist APIs

teng-li committed:
Let FindNCCL to determine the NCCL version

teng-li committed:
Let NCCL2 Backend use ATEN instead deprecated THPP

teng-li commented:
Default build failure seems not to be related to this PR. 
May I have a review? @apaszke @soumith.  Thank you guys!

teng-li commented:
@pytorchbot retest this please

teng-li commented:
[](url)

ngimel commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,740 @@
+
+  if (input.size() <= 0) {
+    throw std::runtime_error("Input tensor sequence cannot be empty");
+  }
```
I don't think it should be a runtime error - this would require special-casing models with no parameters everywhere. IMO, doing something with an empty list of tensors should be a no-op rather than runtime error.

ngimel commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,740 @@
+  }
+  if (input.size() != output.size()) {
+    throw std::runtime_error("Input tensor sequence should have the same size "
+                             "as output tensor sequence");
```
"should have the same number of tensors" I think is more descriptive error message than "same size"

ngimel commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,740 @@
+  THCudaCheck(cudaSetDevice(curDevice));
+
+  // Move into the hash table
+  _ncclCommsAndEvents[groupId] = std::move(std::make_pair(std::move(comms),
```
may be use emplace instead of the first std::move here?

ngimel commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,740 @@
+      throw std::runtime_error("Expecting inputs on different GPU devices");
+    }
+
+    usedDevices.insert(input[i].get_device());
```
insert is making it's own check of whether inserted element is already present in the set, so to avoid doing it twice you can check .insert() return https://stackoverflow.com/a/41152203

csarofeen commented:
Distributed table needs to be modified to enforce ordering of all reduce calls. If the order is not the same across processes NCCL will hang.

teng-li commented:
@csarofeen Could you elaborate more on this?

csarofeen commented:
Since backward is multi-threaded the following could be possible:
processes one finishes params in bucket[0] then params in bucket[1]
processes two finishes params in bucket[1] then params in bucket[0]
The way the code is structured means that process one will try to all_reduce bucket[0] and process two will try to all_reduce bucket[1].
This will deadlock nccl.
Distributed should enforce that all_reduce call on bucket[i] cannot be issued before all_reduce call on  bucket[i-1]

teng-li commented:
@csarofeen Thanks for the explanation, I will get a work-around on this to maintain the thread order.

teng-li committed:
Let distributed parallel model use a single reduction thread for NCCL backend

teng-li commented:
@ngimel Addressed your comments, @csarofeen I made distributed.py use a single reduction thread since all the NCCL calls are asynchronous, using multiple threads wouldn't provide much perf gain for us. Also, it would be pretty tricky to get each thread execute in a certain order.  This simple one thread model works without NCCL deadlock issues as I tested.  So each reduction bucket will write the reduction request to a single queue, which will be call NCCL backend in the order which the gradients become available. The single queue will be assigned to a dedicated single reduction thread to process the NCCL reduction one after another while maintaining the NCCL call order.

teng-li commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,740 @@
+  }
+
+  if (_ncclCommsAndEvents.find(groupId) != _ncclCommsAndEvents.end()) {
+    return std::make_pair(_ncclCommsAndEvents[groupId].first.get(),
```
Because simply returning __ncclCommsAndEvents[groupId] won't compile with implicit conversion.

ngimel commented:
You also have to use gpuGuard.setDevice instead of cudaSetDevice, it won't work otherwise (when you are creating gpuGuard without arguments, it sets original_device to -1, when you are destroying it without actually setting devices in between, it's a no-op).

teng-li commented:
@ngimel Oh right, forgot to update it, now updated.

csarofeen commented:
@teng-li Sorry, I think I'm missing something. I don't understand how you guarantee the order of all_reduce calls will be the same across processes. Can you point me to the right section?

teng-li commented on torch/nn/parallel/distributed.py:
```diff
@@ -253,38 +291,75 @@ def _queue_reduction(self, bucket_idx):
+            # nodes to avoid the deadlock. In other words, we will only maintain
+            # a single reduction thread for this purpose. This is OK since all
+            # distributed NCCL calls are asynchronous
+            self._reduction_queues = [queue.Queue()]
```
@csarofeen I made a single reduction queue and single reduction thread here for NCCL backend.

teng-li commented on torch/nn/parallel/distributed.py:
```diff
@@ -242,7 +274,13 @@ def _queue_reduction(self, bucket_idx):
+
+        if dist._backend == "nccl":
+            # NCCL backend, all buckets will share a single reduction queue
+            self._reduction_queues[0].put((dev_buckets, dev_events, event))
```
@csarofeen Every reduction bucket will write to this single queue when any of the bucket is full (in line 257). Among all processes on different nodes, the order for each bucket to become full is deterministic since the way we divide the gradients into bucket is the same across multiple processes. Therefore, the single queue and single reduction thread makes sure that the order in which each bucket got all_reduced is the same for each process.

csarofeen replied:
"the order for each bucket to become full is deterministic" This is not true. Even though your buckets are made up of all the same gradients, backwards calls are asynchronous which means the order in which the gradients finish are not deterministic. If gradient 3 is submitted before gradient 4 but gradient 3 takes much less time to compute, gradient 4 can finish before gradient 3.
I believe logic needs to be added to enforce the all_reduce of `bucket_idx-1` has been submitted before the all_reduce of `bucket_idx`.

csarofeen replied:
Sorry I take this comment back. You're right. Since autograd is submitting gpu work on a single stream/gpu this should be sufficient. Thank you for making the changes! I will pull these commits and test it out.

apaszke replied:
A nicer way to implement this would be `queue_idx = 0 if _backend == "nccl" else bucket_idx` and then use this to queue

apaszke replied:
Also, I can't see how this makes the order deterministic. Currently, we have no guarantees about the ordering of ops in backward, which means that reduction of 4th block can be queued before 3rd block on one node, while on the other 3rd block will be queued before the 4th (all are queued on this single queue).

teng-li replied:
Since backward goes to the same stream on the GPU, why still no this guarantee?

teng-li commented:
@apaszke Would like to hear your comments too.

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,731 @@
+                             ncclOp[operation],
+                             (*comms)[i],
+                             stream));
+    THCudaCheck(cudaEventRecord((*events)[i], stream));
```
What do we use the events for?

teng-li replied:
Now these events are still needed for destroyGroup and destroyDataChannel functions

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,731 @@
+                                THDGroup groupId) {
+
+  std::vector<at::Tensor> dataVec;
+  dataVec.push_back(data);
```
`std::vector<at::Tensor> data_vec = {data};`

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,731 @@
+    gpuGuard.setDevice(devices[idx++]);
+    THCudaCheck(cudaEventSynchronize(event));
+  }
+}
```
That doesn't really implement a distributed barrier. It only syncs with locally queued kernels

teng-li replied:
It will for NCCL

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.hpp:
```diff
@@ -0,0 +1,173 @@
+
+  private:
+    QueueWorker::Request _request;
+  };
```
I don't think you need to implement anything (no memebers are necessary) for requests if you don't support `send` and `recv` anyway

apaszke commented on torch/nn/parallel/distributed.py:
```diff
@@ -176,6 +187,20 @@ def train(self, mode=True):
         for module in self._module_copies[1:]:
             module.train(mode)
 
+    def recreate_groups(self):
```
Why do you need to recreate groups?

apaszke replied:
I'd rather have us understand all the quirks of NCCL2 before we merge this.

teng-li replied:
Agreeing on understanding this is important. But this shouldn't be the blocker for this PR.  We currently don't have any "working" distributed training solution for PyTorch using Infiniband (The known deadlock). And this workaround at least "works" for now and I have trained imagenet for 120 Epochs on 8 nodes without any issues.

apaszke replied:
It "works", but it makes the code of DDP and its usage much more complicated, which means trouble for all its current users and maintainers. I'm fine with prolonging changes to backends, but this is user visible part of the code, and it's already quite complicated, so I don't want to add more unless there's a good argument for it

teng-li replied:
Ok, let's do this then. I am gonna remove part of these user-facing code in distributed.py. Let's only include the backend code and these new python functions in this PR instead. The distributed.py has already had the comment saying that it supports gloo only.

apaszke commented on torch/nn/parallel/distributed.py:
```diff
@@ -94,8 +94,19 @@ def __init__(self, module, device_ids=None, output_device=None, dim=0):
+            dist.barrier(grp)
+            dist.destroy_group(grp)
+            self.bcast_grp = dist.new_group()
+            self.reduce_grp = set()
```
What's going on in here? Why do you create a temporary group, destroy it, and then create a brand new one? Can't you just reuse the same one? Also, why is `dist.barrier` necessary?

apaszke replied:
Ok, but why? @csarofeen @ngimel any ideas? keeping the group alive would seem correct to me

teng-li replied:
@csarofeen @ngimel Some further context, if grp is not destroyed here, distributed.py will deadlock after the next broadcast in sync_params() function

teng-li replied:
@csarofeen @ngimel Wondering if you guys have any ideas why NCCL is deadlocking in this case? Each group creates the NCCL communicator and CUDA events and they will be cached. Destroy these groups will destroy both of the communicator and events.

if we don't destroy them, then after the first broadcast in distributed.py: __init__, after the next broadcast in distributed: _sync_params, it will deadlock. 

I'm happy to provide further contexts on the workaround

apaszke commented on torch/nn/parallel/distributed.py:
```diff
@@ -187,7 +212,14 @@ def _sync_params(self):
+
+            if dist._backend == "nccl":
+                # TODO: deprecate this broadcast along with
+                # dist.all_reduce_multigpu
```
??

apaszke commented on torch/nn/parallel/distributed.py:
```diff
@@ -187,7 +212,14 @@ def _sync_params(self):
+            if dist._backend == "nccl":
+                # TODO: deprecate this broadcast along with
+                # dist.all_reduce_multigpu
+                dist.broadcast(flat_buffers, 0, self.bcast_grp)
```
do all NCCL calls need an explicitly specified group? why can't you just use the default one like in all other backends?

apaszke replied:
I'm sorry I didn't get that

apaszke commented on torch/nn/parallel/distributed.py:
```diff
@@ -253,38 +291,75 @@ def _queue_reduction(self, bucket_idx):
-                        for reduction_stream in dev_r_streams:
-                            default_stream.wait_stream(reduction_stream)
+                # We sync on the default streams for NCCL backend since all
+                # nccl reduction kernels goes to the default streams
```
Can't you just make `reduction_streams` point to default streams for NCCL? This would allow you to use the same code

apaszke commented on torch/nn/parallel/distributed.py:
```diff
@@ -253,38 +291,75 @@ def _queue_reduction(self, bucket_idx):
-            self._reduction_threads.append(threading.Thread(
-                target=self._reduction_thread_fn,
-                args=(reduction_queue, group_id, self.device_ids, reduction_streams, self._nccl_streams)))
+            if dist._backend == "nccl":
```
you have three conditionals for NCCL in a row. Please concat them to form a single large conditional

apaszke commented on torch/nn/parallel/distributed.py:
```diff
@@ -320,3 +395,47 @@ def _process_batch():
+
+            # TODO: remove nccl.reduce with
+            #       dist.all_reduce_multigpus
+            nccl.reduce(dev_coalesced, root=0, streams=default_streams)
```
This is confusing. You have already implemented the multigpu calls, so why not use the,?

teng-li replied:
Will do it in a separate PR.  if we use allreduce_multigpu,  then user needs to use optimizer.step()  to update the gradient on each of the 8 ( or any number) GPUs instead of 1. In other words, passing each GPU's module's parameter to the optimizer and then do the step() for eight times.

This process is currently the performance bottleneck and I am still trying to figure out how to make it faster.  It would be great if you could comment on this as well

apaszke replied:
that doesn't seem right - the parameters are broadcasted to all GPUs in `_sync_params`, which is why it's enough to use a single optimizer.

I'm worried about adding these new functions in these PR. They have no tests and are not used in here either. There's no way for us to figure out if they're correct.

teng-li replied:
Once I move to dist. allreduce_multigpu, why do you still think we need to do broadcast parameters in _sync_params? Aren't we supposed to delete it? 

allreduce_multigpu should will let every single GPU to have the averaged gradients. And all the rest left to do at the beginning of each forward is to update the parameters on every GPU with the optimizer on 8 GPUs.

So a desired flow is that we don't broadcast anything in forward (maybe buffer?, but not parameters), and at the reduction thread, only use a single step of dist.all_reduce_multigpu(), which will use all 4 IB interfaces.  

For these new functions, I have my own script of testing it and I have already tested the functionality of the new functions.  As I put in the TODO in this PR, I will write the unittest for this too later.

apaszke replied:
I know this is possible with NCCL, but this changes the public interface - the user would have to have separate optimizers for everything in `_module_copies`, and the copies are currently private. This is a breaking change of the interface.

If you don't want to unit test these functions, the please at least use them in Python code and check that it runs. Additionally, they look like ordinary non-experimental distributed functions, so please make them raise a warning saying that they are and shouldn't be used in user code (or just prefix them with an underscore).

teng-li replied:
Yeah, that's why I didn't change the dist.all_reduce to dist.all_reduce_multigpu right now since it involves these interface changes and we later need to get some cleaner ways of doing it.

For the testing, yes, that's what I have already done in my own testing script, which is the python code that test all the new functions.  All the four functions work fine from my testing code.

apaszke reviewed (commented):
I don't think requiring users to call `torch.distributed.destroy_process_group()` is necessary. If NCCL needs this then register an atexit hendler.

teng-li commented on test/test_distributed.py:
```diff
@@ -574,6 +574,13 @@ def _join_and_reduce(self, fn):
 
+elif BACKEND == 'nccl':
+    dist.init_process_group(init_method=INIT_METHOD, backend='nccl')
+    # TODO
```
This is a unittest TODO

apaszke replied:
But what does it mean? Do you want to add any tests? Do you want to add more tests? A description would be helpful

ngimel commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,731 @@
+    // Contiguous verification
+    if (!input[i].is_contiguous() || !output[i].is_contiguous()) {
+      throw std::runtime_error("Expecting all GPU tensors to be contiguous");
+    }
```
e.g. distributed all_reduce is done in-place, so making tensor contiguous behind user back is somewhat risky?

apaszke replied:
Right. We have `freeCopyTo`, but let's just raise an error for now

teng-li commented on torch/csrc/distributed/Module.cpp:
```diff
@@ -338,6 +367,244 @@ PyObject* THDPModule_recv(PyObject *_unused, PyObject *args)
+  std::vector<at::Tensor> descriptors;
+  std::vector<at::Tensor> raw_descriptors;
+  THDGroup group;
+  THDReduceOp op;
```
These predeclarations are required due to the following goto statements

teng-li committed:
Caching the sockets, bug fix, refactoring, and addressed Adam's comments

pietern commented:
@pytorchbot retest this please

apaszke commented on torch/distributed/__init__.py:
```diff
@@ -244,7 +381,10 @@ def all_gather(tensor_list, tensor, group=group.WORLD):
+    if _backend != "nccl":
+        return torch._C._dist_all_gather(tensor_list, tensor, group)
+    else:
+        return all_gather_multigpu(tensor_list, list[tensor], group)
```
Can you please point me to the line that does it? It's weird that one of these lists is expected to be nested deeper than the other. Still, `list[tensor]` is incorrect (you can't index a class)

teng-li replied:
Actually, this needs to be change to [tensor_list], misunderstood your question initially.

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,731 @@
+    for (auto& event : *(itemPair.second.second)) {
+      gpuGuard.setDevice(devices[idx++]);
+      THCudaCheck(cudaEventSynchronize(event));
+      THCudaCheck(cudaEventDestroy(event));
```
I think the point of these changes would be to make the PR smaller and simpler to review. But we can fix that later if you don't want to do it now

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,731 @@
+
+// Use the socket to broadcast NCCL ID
+void DataChannelNccl::broadcastUniqueNcclId(ncclUniqueId* srcNcclId,
+                                            ncclUniqueId* dstNcclId) {
```
But is it even used? It seems that only the first process broadcasts it, all other processes just ignore that argument.

apaszke replied:
I just realized that you're always passing the same address twice, because it's only used for one of those purposes. Why not take in a single pointer? It would also let you call `ncclGetUniqueId` once inside this method only in first process (we discard all other values).

teng-li replied:
Taking a single pointer is fine too.

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,734 @@
+    gpuGuard.setDevice(devices[idx++]);
+    THCudaCheck(cudaEventSynchronize(event));
+  }
+}
```
As I said before that's not a distributed barrier. E.g. if you never do any NCCL calls, it will be a no-op (and barriers are never no-ops!). Consider this code:
```python
print("a")
dist.barrier()
print("b")
```
If you used MPI, you would have a guarantee that all processes will print a before any will print b. This function doesn't have the same guarantees.

I don't think you need to keep these events. Just make these functions always queue work on whatever is current stream, and keep the synchronization up to the caller (just like we do in Gloo channel)

teng-li replied:
Yes, this is not a general-purpose barrier, it's just a barrier for NCCL. I still think this function would be a good option for the user and would rather not let user to simple to cudaStreamSynchronize on the default stream if they ever need to wait.  I am gonna make this function a new function, something called ncclBarrier()

If you don't

apaszke replied:
It's not, but semantically that's what it is supposed to mean. Not raising an error means breaking the contract of the data channel interface which is a bad thing. I think stream synchronization shouldn't be part of DataChannel. It's up to the user to ensure that streams they use wait for other streams (maybe they will want to sync with a stream different than the default one?). We already have code for this in DistributedDataParallel, so it doesn't add much

teng-li replied:
Raise an error instead in the new code change

teng-li committed:
Make BcastNcclID take a single param and bug fix for all_gather

teng-li committed:
Removed barrier function, added warning for users, and not exposing experimental func to users

teng-li commented:
@apaszke addressed your comments: (1) removed barrier support, but we still need these events for destroy functions, and later these events are useful too for benchmarking and other extension work, so I keep them in the backend code. (2) added warning for users to use with care with NCCL2, (3) made the experimental function in distributed.py private in order to not confuse the user.

csarofeen commented:
@teng-li Can you create a small pytorch repro script for the deadlocking outside data parallel? If so I can take a look at it and send it over to the NCCL guys to check as well.

teng-li committed:
Use the simplest single bucket working solution for distriubted data parallel model with rebase

teng-li commented:
@csarofeen Will do.
@apaszke Made the simplest change to distributed.py with a single thread, which actually gives the best performance on my testing. Now distributed data parallel model works with NCCL2 backend.

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,715 @@
+  if (ncclDatatype.find(type) == ncclDatatype.end()) {
+    throw std::runtime_error("Unsupported data type for NCCL");
+  }
+  return ncclDatatype[type];
```
nit: `find` gives you an iterator that will give you the answer immediately.
I think it's nicer to rewrite this as that:
```cpp
try {
  return ncclDatatype.at(type);
} catch (std::out_of_range& e) {
  throw std::runtime_error("Unsupported data type for NCCL backend");
}
```

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.hpp:
```diff
@@ -0,0 +1,232 @@
+  // Socket on which the master is listening
+  int _masterListeningSocket;
+  // Sockets on which the master is sending to each slave
+  std::vector<int> _masterSendingSockets;
```
It would be useful to add a comment that these are not sorted by ranks and can be in arbitrary order

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,715 @@
+    // Destroy the CUDA events
+    size_t idx = 0;
+    for (auto& event : *(_groupNcclResources[groupId].ncclCudaEvents())) {
+      auto devices = getDevicesList(_groupDevices[groupId]);
```
This doesn't need to be inside this loop and could be computed only once

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,715 @@
+      NCCL_CHECK(ncclCommDestroy(comm));
+    }
+    _groupNcclResources.erase(groupId);
+  }
```
Would be nice if we could share this code with `destroy()`. Otherwise it's easy to miss something in one of these places when you'll change the other one

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,715 @@
+  // Will keep the default group and only destroy its CUDA resources.
+  if (groupId != THDGroupWORLD &&
+      _groups.find(groupId) != _groups.end()) {
+    _groups.erase(groupId);
```
That's quite a weird edge case. I'd rather have `destroyGroup` actually destroy the group *unconditionally*.

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,715 @@
+    _groups.erase(groupId);
+  }
+  if (_groupDevices.find(groupId) != _groupDevices.end()) {
+    _groupDevices.erase(groupId);
```
`erase()` can be called even when this key is not in the map

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,715 @@
+  // Now creating the CUDA events
+  for (size_t i = 0; i < input.size(); ++i) {
+    gpuGuard.setDevice(input[i].get_device());
+    THCudaCheck(cudaEventCreate(&((*events)[i])));
```
`&((*events)[i])` is exactly why I'd prefer to keep vectors as plain members :wink:

apaszke commented on torch/lib/THD/base/DataChannel.hpp:
```diff
@@ -75,20 +75,52 @@ struct DataChannel {
+                       rank_type src_rank,
+                       THDGroup group_id = THDGroupWORLD) = 0;
+  // All reduce multiple GPUs on a number of nodes
+  virtual void allReduce(std::vector<at::Tensor>& input,
```
I'm fine with extending the API, but this goes against the single-arg method because it accepts outputs that can be different than inputs. Can you please make them work in the same way (in-place)?

apaszke commented on torch/lib/THD/base/DataChannel.hpp:
```diff
@@ -75,20 +75,52 @@ struct DataChannel {
-                      rank_type dst_rank, THDGroup group_id = THDGroupWORLD) = 0;
-  virtual void broadcast(at::Tensor& data, rank_type src_rank,
+  // Reduce multiple GPUs on a number of nodes
+  virtual void reduce(std::vector<at::Tensor>& data,
```
This doesn't specify which tensor in `dstRank` will hold the result. Is it the first one?

apaszke commented on torch/lib/THD/base/DataChannel.hpp:
```diff
@@ -75,20 +75,52 @@ struct DataChannel {
+                      THDReduceOp operation,
+                      rank_type dst_rank,
+                      THDGroup group_id = THDGroupWORLD) = 0;
+  // Reduce multiple GPUs on a number of nodes
```
nit: not reduce

apaszke commented on torch/lib/THD/base/DataChannel.hpp:
```diff
@@ -75,20 +75,52 @@ struct DataChannel {
-  virtual void allGather(std::vector<at::Tensor>& output, at::Tensor& input,
+  // All gather multiple GPUs on a number of nodes
+  virtual void allGather(std::vector<at::Tensor>& input,
+                         std::vector<at::Tensor>& output,
```
This signature is also weird. Firstly, it swaps output and input compared to the other overload. Secondly, I'd expect that if you get a vector of inputs, you also get a vector of outputs (a vector per device).

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,715 @@
+  }
+  NCCL_CHECK(ncclGroupEnd());
+
+  cudaFreeMutexLock.unlock();
```
It will be unlocked automatically.

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.cpp:
```diff
@@ -0,0 +1,715 @@
+  THDGroup newGroupId = static_cast<THDGroup>(_groups.size());
+
+  // Insert the current group
+  _groups.insert({
```
It seems that you're not using `_groups` for anything useful. Why even bother with maintaining them?

apaszke replied:
I guess? Keeping a single counter would be a lot simpler than maintaining a whole map that's unused.

teng-li committed:
Cleanup, fixes and further addressed Adam's comments

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.hpp:
```diff
@@ -0,0 +1,232 @@
+  private:
+
+    std::pair<std::unique_ptr<std::vector<ncclComm_t>>,
+              std::unique_ptr<std::vector<cudaEvent_t>>> _commEventPair;
```
They will always be very small, so any costs you could possibly pay because of missing a move (I'm also looking for missed moves when reviewing), are probably outweighed by the cost of double indirection when accessing memory all the time. Storing them as two members is going to make the code a lot clearer - if you didn't make them private, you could just access them like regular fields, and wouldn't need `NcclResourcePair` at all - you could just pass `NcclResources&` or `NcclResources*` around and have everything work just fine.

apaszke commented on torch/lib/THD/base/data_channels/DataChannelNccl.hpp:
```diff
@@ -0,0 +1,232 @@
+   *      Note that the order of the device for the tensor list matters and
+   *      each group can only be associated with one used device string
+   */
+  std::unordered_map<THDGroup, std::string> _groupDevices;
```
The problem with this approach is that it doesn't really fit the interface we expose - a group is just a way to have multiple collectives running concurrently. I think a better solution, would be to store a communicator per group and per combination of devices used within this group. That would be equivalent and would satisfy the interface, without resorting to clearing caches, etc. It's not like this will create any more groups than if you didn't do this - it's just that certain communicators will belong to a single group instead of multiple.

I'm fine with merging this, but we definitely won't make this backend public until it's fixed.

Do we finally know what's causing the deadlocks and why do we even need to destroy communicators periodically?

teng-li replied:
I don't know the exact reason why still but this is pretty much the fix for this symptom and that's something I would like to further investigate as well. Then I will probably open up a new PR to address this issue as well, by not associating a given process_group with just a set of GPU orders. I mean, I am not a big fan of caching all the communicators, I am thinking of caching probably the latest one, and also cache all the NCCL unique ID to avoid the broadcast. This is more like an optimization code and I will also do in a separate PR.

teng-li replied:
Updates, now this is fixed and works. Instead of storing multiple communicators per group per device sequence. We only cache one communicator, if a new device sequence comes in, we destroy the previous communicator.  (The device sequence change shouldn't happen that often anyway in a training process)

apaszke commented on torch/nn/parallel/distributed.py:
```diff
@@ -22,7 +22,7 @@
 
 class DistributedDataParallel(Module):
-    r"""Implements distributed data parallelism at the module level.
+    """Implements distributed data parallelism at the module level.
```
Can't see it. You might have forgotten to check in some changes

teng-li replied:
Probably forgot to check in the changes, done

apaszke reviewed (commented):
The DataChannel itself could be simplified, but that's not a blocker for this merge (only for public release). There are two things that need fixing - the raw string prefix in DDP docstring, and incorrect usage of `PySequence_Fast`.

teng-li committed:
Used PySequence_Fast in distributed csrc

teng-li commented on torch/csrc/distributed/Module.cpp:
```diff
@@ -338,6 +367,254 @@ PyObject* THDPModule_recv(PyObject *_unused, PyObject *args)
+    goto invalid_arguments;
+  }
+
+  tmp_length = PySequence_Fast_GET_SIZE(sequence);
```
Oh, yeah, wrote too fast yesterday to check. Fixed now and thanks for pointing out

teng-li commented on torch/distributed/__init__.py:
```diff
@@ -51,6 +73,31 @@ def init_process_group(backend, init_method='env://', **kwargs):
+    elif backend == "nccl":
+        _backend = dist_backend.NCCL
+    else:
+        raise RuntimeError("Invalid distributed backend name detected")
```
Will do

teng-li commented on torch/distributed/__init__.py:
```diff
@@ -85,7 +80,11 @@ def init_process_group(backend, init_method='env://', **kwargs):
         _backend = dist_backend.NCCL
     else:
-        raise RuntimeError("Invalid distributed backend name detected")
+        raise RuntimeError("Invalid distributed backend name: " + backend)
```
@apaszke  Did this change so that the wrong backend string won't be passed into the torch._C function. We will get an error like this.

```
tengli@learnfairXXXX:~$ python -u imagenet/main.py -j 16 -a resnet50 --print-freq 1 --batch-size 256 --world-size 2 --dist-url 'tcp://XXX.XXX.XXX.XXX:20000' /datasets/XXXX/XXXX/ --epochs 10 --dist-backend 'abc' --rank 0
Traceback (most recent call last):
  File "/private/home/tengli/pytorch_examples/imagenet/main_distributed.py", line 371, in <module>
    main()
  File "/private/home/tengli/pytorch_examples/imagenet/main_distributed.py", line 78, in main
    rank=args.rank)
  File "/private/home/tengli/miniconda3/lib/python3.6/site-packages/torch/distributed/__init__.py", line 83, in init_process_group
    raise RuntimeError("Invalid distributed backend name: " + backend)
RuntimeError: Invalid distributed backend name: abc
```

teng-li committed:
Removed the limitation that each group is only bound to a given device sequence

teng-li commented:
@apaszke  

Added two commits: The first one used the correct PySequence Fast for everyone in the module.cpp and added the "r" back. Also changed and tested and made sure that the invalid backend error can be thrown.

The second one removed the limitation that each group is bound to a given device sequence. But I don't feel comfortable to cache many NCCL communicators. So the current design is, for each process group, we only cache a single NCCL communicator (say with device sequence 0,1,2,3,4,5,6,7) If there is a new collective NCCL call with a new device sequence (say 1,0,2,3,4,5,6,7). We will destroy the existing cached NCCL communicator (with sequence 0,1,2,3,4,5,6,7), rebuild a new communicator and cache this one instead. This design is pretty clean and simple. I tested with different new device sequence and made sure it worked. This shouldn't be a release blocker for now

For the NCCL deadlock, it's not super clear to me why but it happens on several consecutive broadcast calls using the same communicator. If the communicator is rebuilt, it will work (as the current workaround).

apaszke commented on torch/csrc/distributed/Module.cpp:
```diff
@@ -338,6 +367,271 @@ PyObject* THDPModule_recv(PyObject *_unused, PyObject *args)
+    goto invalid_arguments;
+  }
+
+  sequence = PySequence_Fast(PyTuple_GET_ITEM(args, 0), nullptr);
```
This returns a new reference, so you will leak memory if you don't decref it. That's why I suggested using `THPObjectPtr`, because it will do it for you automatically (even in case of any errors).

teng-li replied:
This is now fixed in the new commit

apaszke commented on torch/csrc/distributed/Module.cpp:
```diff
@@ -338,6 +367,271 @@ PyObject* THDPModule_recv(PyObject *_unused, PyObject *args)
+  }
+
+  tmp_length = PySequence_Fast_GET_SIZE(sequence);
+  THPUtils_assert(tmp_length >= 0, "couldn't obtain the length of %s",
```
nit: this cannot fail now, so no need for error checking. You can safely assign to `length` directly.

apaszke commented on torch/csrc/distributed/Module.cpp:
```diff
@@ -338,6 +367,271 @@ PyObject* THDPModule_recv(PyObject *_unused, PyObject *args)
+    goto invalid_arguments;
+  }
+
+  sequence = PySequence_Fast(PyTuple_GET_ITEM(args, 0), nullptr);
```
I'm not 100% sure if error message can be NULL. The docs don't say that it can, so I'd assume it shouldn't. Just put a simple error message saying that this function expected a sequence.

apaszke reviewed (commented):
Rebuilding communicators sounds quite expensive, so I'm not sure if that's the best solution, but let's stick to that for now.

I still have a few comments for the Python bindings, but that's it. Right now they are leaking memory.

apaszke commented:
Also, when you make new changes, please add them as new commits at the end instead of squashing. It makes repeating the review much easier, because I explicitly see what changed. You can always clean up the history right before merge.

teng-li committed:
Used THPObjectPtr for PySequence_Fast

seba-1511 commented:
@teng-li This work looks quite good. I like the distributed interface much more than the NCCL one, even for single machine training.

Is there a way I can contact you privately ? I would like to start playing with this as soon as possible, so as to add it to the distributed tutorial.

teng-li commented:
@seba-1511 Thank you, sure, you can leave me your email and I will send you an email.

seba-1511 commented:
My email is: arnolds _at_ usc _dot_ edu
